[SYCL][Doc] Add InvokeSIMD and Uniform extensions

Pennycook opened the pull request:
These extensions introduce a mechanism for asserting that the value of a variable is uniform, and support for an execution model in which SIMD parallelism is expressed using SIMD types, with clear delineation between SPMD and SIMD code regions through an invocation interface.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

gmlueck commented:
There should be a feature-test macro name for this extension, as described in the SYCL 2020 guidelines for extensions (https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_feature_test_macros).  For example, we might choose `SYCL_EXT_ONEAPI_INVOKE_SIMD`.

Since we can use the feature-test macro name as a way to identify this extension, I think there is no need for these extension specs to have a separate "name string".  Therefore, I think we can remove the description of the name `SYCL_INTEL_invoke_simd`.

Pennycook commented:
> There should be a feature-test macro name for this extension, as described in the SYCL 2020 guidelines for extensions (https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_feature_test_macros). For example, we might choose `SYCL_EXT_ONEAPI_INVOKE_SIMD`.
> 
> Since we can use the feature-test macro name as a way to identify this extension, I think there is no need for these extension specs to have a separate "name string". Therefore, I think we can remove the description of the name `SYCL_INTEL_invoke_simd`.

Thanks, @gmlueck.  I've made those changes in 83b3319.  I also renamed the extension (and the asciidoc file) to match the new feature test macro.

keryell commented:
@jbrodman is this `ispc` inside a sub-group? :-)

AlexeySachkov commented on sycl/doc/extensions/InvokeSIMD/SYCL_INTEL_invoke_simd.asciidoc:
```diff
@@ -0,0 +1,364 @@
+  becomes a +std::tuple<float, int>+).
+
+Note that these conversion rules do not permit user-defined types or special
+types (e.g. +nd_item+) being passed between SPMD and SIMD contexts.  This
```
I'm not sure that this is fully enforced, because as stated above:
> Each argument in the +args+ parameter pack must be an arithmetic type, a
> trivially copyable type wrapped in a +sycl::ext::oneapi::uniform+, or a
> +std::tuple+ of these types.

It is possible to pass a user-defined trivially copyable type to `uniform` wrapper and therefor make it cross the border between those two contexts

Pennycook replied:
Good catch.  I think this should say something like the below, instead:
> Note that special SYCL types (e.g. +nd_item+) cannot be passed between SPMD and SIMD contexts.

@rolandschulz, do you agree?  Should we go one step further and disallow the creation of `uniform<T>` objects if `T` is a special type?

jasonsewall-intel replied:
I think we'd have to carefully spell out what special types we're talking about, but it makes sense that something that specifically is intended to vary across work-items shouldn't be able to passed to uniform. 

I would hope that the compiler would scream about this anyway, but making it formally an error is probably more responsible. It's interesting to think of what this means in the context of SG size 1, but I don't think that's a special case worth making exceptions for.

Pennycook replied:
> I think we'd have to carefully spell out what special types we're talking about, but it makes sense that something that specifically is intended to vary across work-items shouldn't be able to passed to uniform.

Agreed.  Added this in c1a865ef8d66.

> It's interesting to think of what this means in the context of SG size 1, but I don't think that's a special case worth making exceptions for.

Yeah.  We also can't detect whether the sub-group size is 1 or not at (host) compile-time, so it could be difficult to special case.

AlexeySachkov commented on sycl/doc/extensions/InvokeSIMD/SYCL_INTEL_invoke_simd.asciidoc:
```diff
@@ -0,0 +1,364 @@
+- Mapping SIMD operations to an unrolled loop executed by a single work-item.
+
+Since this execution model guarantees SIMD-like behavior, there is no need for
+the user to insert sub-group barriers in an explicit SIMD function.
```
I guess this is also impossible, because you can only obtain a `sub_group` handle through `nd_item`, which can't be easily passed down to `SIMD` context (because it is not an arithmetic type and it is not uniform across sub-group) and that handle is required as the first argument of `group_barrier` free function - right?

Pennycook replied:
That's right.  The intent of this paragraph was to make it clear to users that they no longer have to worry about barriers, and to make it clear to implementers that if their implementation requires barriers between statements then the user shouldn't be responsible for generating them.  Would the below be clearer:
> Since this execution model guarantees SIMD-like behavior, there is no need for the user to insert any form of explicit synchronization functions to ensure memory consistency across SIMD lanes.

AlexeySachkov replied:
> Would the below be clearer:

The existing wording is fine by me, I just wanted to double-check my understanding and my comment could also be treated as a "prequel"  to [which SYCL APIs can be used within SIMD context?](https://github.com/intel/llvm/pull/3445#discussion_r607864083) question asked below

Pennycook replied:
I've changed the wording anyway (in 18f3232b186e).  Answering your question made me realize that mentioning sub-group barriers here is unnecessary, and I'm worried that it could lead to confusion.

AlexeySachkov commented on sycl/doc/extensions/InvokeSIMD/SYCL_INTEL_invoke_simd.asciidoc:
```diff
@@ -0,0 +1,364 @@
+
+Uniform arguments are represented by scalars, and varying arguments by explicit
+vector types.  The explicit vector are types are based on the SIMD classes
+from [Parallelism TS2](http://wg21.link/N4808).
```
It is not clear to me what are exact things which are imported from Parallelism TS2 and which of them are allowed to be used in SYCL SIMD functions.

I see that `std::experimental::where` usage is demonstrated in examples below, but are we going to allow the rest of functions to be also used, like `std::experimental::min`? What about math functions? Are we plan to provide functionality like `permute` or `exclusive_scan` for `sycl::ext::oneapi::simd`/`std::experimental::simd`?

Pennycook replied:
After discussing this internally we've decided that we should be aiming to implement all of the SIMD-related functions from Parallelism TS 2.

Pennycook replied:
I've tried to clarify this in 53a40f3fd670; please let me know if there are still cases that you're unsure about.

Nothing explicitly states that a function like `sycl::exclusive_scan` cannot be called in a SIMD region, but I think that's covered by the fact that there's no way to pass a group object into the SIMD region.

Pennycook commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,390 @@
+. Should arguments require trivially copyable or device copyable?
++
+--
+*UNRESOLVED*: Device copyable permits bitwise copies of a wider range of classes, but the
+SYCL 2020 specification states that this only applies to inter-device transfers.
```
I'd be interested to hear from @rolandschulz and @nliber on this.

The paragraphs from SYCL 2020 that this extension is referencing are these two (from https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable):
> The SYCL implementation may need to copy data between the host and a device or between two devices
> ...
> When doing an inter-device transfer of an object of type `T`, the effect of each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is the same as a bitwise copy of the object;

When the specification says "inter-device" here, I think it's classifying the host as a device.  I agree that something like "a transfer between host and device or between two devices" would by clearer.

The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary.  An implementation may still execute constructors and copy operators in the device code.

AlexeySachkov replied:
> The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary.  An implementation may still execute constructors and copy operators in the device code.

Is it possible for user to write two different constructors: one which is invoked during host-device communication and another one which is invoked only on device side? I guess it should be possible with `__SYCL_DEVICE_ONLY__`. Do we really want to support that? Is there any particular use case where two versions of constructors are needed?

Pennycook replied:
> Is it possible for user to write two different constructors: one which is invoked during host-device communication and another one which is invoked only on device side? I guess it should be possible with __SYCL_DEVICE_ONLY__. Do we really want to support that? Is there any particular use case where two versions of constructors are needed?

I think that would be an ODR violation, and so wouldn't be allowed.

The "When doing an inter-device transfer of an object of type `T`" is what's confusing me, here, but you're right -- if there can only be one definition of each of those constructors and operators, it's unclear how a user could specialize the functions for that use-case.  But I'm going to stop guessing, and wait for the experts (@nliber and @rolandschulz) to weigh in.

rolandschulz replied:
> I think that would be an ODR violation, and so wouldn't be allowed.

ODR violation apply to a single compiler. We aren't saying that this isn't allowed for split compilation (but we also not clearly saying it is allowed).

> The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary. An implementation may still execute constructors and copy operators in the device code.

I agree this is a correct reading of the wording. I don't think we spent any time wondering whether it should apply to more than device transfer because that was the use-case this was added for.

I'm not aware of any use case of device-copyable for which the copy constructor/assignment on the device does anything different from bitwise copy. And if we changed the requirement for the user to declare a type device-copyable that it doesn't do anything extra on the device we could use it here. But I think it would require a change of that paragraph to clarify it.

Pennycook replied:
Thanks, Roland.

@AlexeySachkov: Given Roland's answer, I propose that we merge this extension specification with the question about `is_device_copyable` unresolved.  I'll open a Khronos issue to discuss the intent of the specification and the need for potential clarification, and we can revisit this later.  Does that work for you?

AlexeySachkov replied:
> @AlexeySachkov: Given Roland's answer, I propose that we merge this extension specification with the question about is_device_copyable unresolved. I'll open a Khronos issue to discuss the intent of the specification and the need for potential clarification, and we can revisit this later. Does that work for you?

Sure, no objections

kbobrovs commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,390 @@
+namespace oneapi {
+
+template <typename Callable, typename...T>
+auto invoke_simd(sub_group sg, Callable&& f, T... args);
```
(repeating the Q from internal discussion here)

Did you have some particular implementation approach in mind?

I’m thinking that invoke_simd will be an intrinsic which will capture the callable, arguments and argument/ret uniformity information (not 100% sure if it is needed). So that SPMD BE could do proper transformation and replace the call to invoke_simd with actual call after widening the caller.

If you had something similar in mind, then Ideally this would be a SPIRV intrinsic – is there a plan to standardize this part?

Pennycook replied:
I suspect that our implementation will look very similar to what you've described, at least for the backends that consume SPIR-V.  We need _something_ at the SPIR-V level to convey that we're transitioning from a SPMD context to a SIMD context, and a special `OpFunctionCall` seems like a good way to do that.  If we introduce such a thing, I think we should standardize it.

I don't think this specification should depend on it, though.  There are other valid implementations, and backends that don't go through SPIR-V should have the freedom to do something else.

kbobrovs replied:
Good, thanks.

> I don't think this specification should depend on it, though.

Sure, agreed.

kbobrovs reviewed (approved):
LGTM

Pennycook commented:
We received feedback that it wasn't clear how or why the `sycl::ext::oneapi::uniform` wrapper class could be used outside of calls to `invoke_simd`.  To remedy this, I've split out all the text discussing `uniform` into a separate `SYCL_EXT_ONEAPI_UNIFORM` extension, and added some simple usage examples in a non-normative section.

@gmlueck, could you please take a look?  Also pinging the other @intel/dpcpp-specification-reviewers -- discussion seems to have stalled on the `is_device_copyable` question raised in https://github.com/intel/llvm/pull/3445#discussion_r609491795; should we be trying to merge this in now, or would people rather we wait until that question has been resolved?

skozhukh commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,365 @@
+sub-group size of the kernel; calling a SIMD function with mismatched lengths
+is an error, and an implementation must issue a diagnostic.
+
+To enable overloading in cases where the `Callable` accepts either a scalar
+type or a SIMD type, we leverage the `sycl::ext::oneapi::uniform` wrapper
+type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.
```
Even we discussed this in email, I would like to state my opinion here as I'm still not convinced by our discussion.  I believe having this special sycl::ext::oneapi::uniform is not needed in general. The compiler can match several possible callees by matching their data type, taking the "underlying" data type from vectors, so f(int ) and f(simd<int>) will both match. And then generate the call to the variant selected by uniformity analysis.  That strategy is perfectly aligned with our function pointers.

Pennycook replied:
We don't want the compiler to do the matching, and we don't want the compiler to do any uniformity analysis.  The purpose of `uniform` is to give developers complete control over matching (using regular C++ overload rules), and to let developers assert uniformity in situations where the compiler can't possibly prove it.

skozhukh replied:
The first statement is a little unclear. Who then will do matching?.. The suggested way just simplifies that matching.  
I'm not against this simplification, but I really don't like a "fully autonomous" uniform specification that can appear anywhere in the program.  That can lead to very hard-detectible mistakes, when user thinks the value is uniform but really it's not. E.g. a function parameter. When a user creates a function and uses it himself it may be always uniform.  But that can change when some one else uses that function.  I'll leave a comment in the uniform specification.

Pennycook replied:
Everything is handled by the C++ overloading mechanism and the type system.  A developer cannot accidentally declare something to be `uniform` by calling a function -- they have to call the `uniform` constructor explicitly.

Here's an example that might help:

```c++
// Function with varying (T) and uniform (uniform<T>) overloads
void foo(int x);
void foo(uniform<int> x);

// Function with varying (T) argument only
void bar(int x);

// Function with uniform (uniform<T>) argument only
void baz(uniform<T> x);

int main()
{
    int x;                // x is assumed varying
    foo(x);               // calls foo(int)
    foo(uniform<int>(x)); // calls foo(uniform<int>)

    bar(x);               // calls bar(int)
    bar(uniform<int>(x)); // calls bar(int) -- uniform<int> converts implicitly to int

    baz(x);               // compiler error -- int does not convert implicitly to uniform<int>
    baz(uniform<int>(x)); // calls baz(int)
}

skozhukh replied:
I meant the code inside the function. E.g.
void foo(int x) {
  int z = uniform<int>(x);
}
User can think that foo is called always with uniform parameter and assert that.

Pennycook replied:
This extension doesn't introduce that problem -- it attempts to solve it!

Today, if a parameter to a SYCL function must be uniform, that must be expressed in the function's documentation.  There is no way to describe via a function's signature that a given parameter must be uniform.  So, it's already possible to write functions that rely on assumed uniformity, like this:
```c++
void foo(sub_group sg, int x) {
  if (x == 5) {
    int z = reduce_over_group(sg, x);
  } else {
    ...
  }
}
```
...if `x` is not uniform, then it is illegal to call `reduce_over_group` inside of the conditional.  The user can document that `foo` requires `x` to be uniform, but cannot enforce it.  If they want to support both uniform- and non-uniform data, the only way to do so is via a run-time check:
```c++
void foo(sub_group sg, int x) {
  if (all_of_group(sg, x == 5)) {
    int z = reduce_over_group(sg, x);
  } else {
    ...
  }
}
```

This extension allows for the uniformity requirement to be declared as part of the function signature.  I don't think people would write the code from your example; they'd write the below instead:
```c++
void foo(uniform<int> x) {
  int z = x;
}
```

skozhukh replied:
Creating a spec should suppose some implementation. I would like to know how you see the implementation for IA.

skozhukh replied:
And one more question. How is this aligned with our function pointers specification? As far as I remember, we had the same intent,  explicit uniform/linear/varying specification for formal and actual parameters. Do you remember why that was not accepted? Are we going to change that spec too?

skozhukh commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,365 @@
+- Arguments of type `sycl::ext::oneapi::uniform<T>` are converted to type `T`.
+  The value may be taken from any work-item in the sub-group; wrapping an
+  argument in a `sycl::ext::oneapi::uniform` is an assertion that the variable
+  holds the same value for all work-items in the sub-group. If this assertion
+  does not hold, the value of the scalar variable passed to the SIMD function
+  is undefined.
```
Is there any way to indicate that for user?

Pennycook replied:
I'm not sure what you mean.  The user can already insert a run-time check for uniformity (using the `all_of` function) prior to taking a branch that asserts uniformity if they are concerned about correctness.

Compilers are encouraged to issue a diagnostic if a variable can _never_ be uniform, but they can't reason about the general case.  In the general case (e.g. a value read from memory), only the developer knows if a given variable is expected to hold the same value across all work-items.

skozhukh commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,365 @@
+  work-item with sub-group local ID `i`.
+
+- Arguments of type `sycl::ext::oneapi::uniform<T>` are converted to type `T`.
+  The value may be taken from any work-item in the sub-group; wrapping an
```
This needs to be clarified for cases when the call is under divergent condition. Calling the function under divergent condition means that not all work items are valid.

Pennycook replied:
Good point.

I think we should update this to say that a `uniform` can only be constructed in converged control flow (at least for now).  All of the use-cases we've seen for this tie into sub-group functions that already have this requirement.

Pennycook replied:
Fixed in d47e0094de20c26dfb797401a9819cb8a9dfa83b.

skozhukh commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,365 @@
+After this mapping has taken place, an appropriate variant of `Callable` is
+selected based on standard overload resolution.
+
+The return value of an `invoke_simd` function is subject to the same mapping
+rules in reverse:
```
I think this needs clarification for calls under condition. Are all elements considered as defined? Or only unmasked ones?

Pennycook replied:
The specification already says that `invoke_simd` can only be called in converged control flow.  Doesn't that cover this case?

skozhukh replied:
It seems I've missed that. Thanks.

skozhukh commented on sycl/doc/extensions/Uniform/Uniform.asciidoc:
```diff
@@ -0,0 +1,190 @@
+
+These concepts are already exposed in other SPMD programming models:
+
+- OpenMP's `declare simd` directive has a `uniform` clause
+- ISPC's `uniform` qualifier makes uniformity part of the type system
```
You can't simply say "this is uniform", something is needed to ensure that in compile-time, or at least at run-time. E.g. "uniform" in OpenMP is ensured by that it will be never used in context when the function parameter is non-uniform.  In ISPC it's a base property of the object, it's created as uniform. 
Just creating uniform from anything can lead to hard run-time errors when "uniform" is used incorrectly. Do we want users to create such a checks at run time? How is that possible? Or are we going to provide some automatic functions?

Pennycook replied:
OpenMP and ISPC are only mentioned here to highlight that the concept of uniformity is important; we're not trying to recreate their semantics in SYCL.  They're very different languages, and the solutions they've adopted don't work for SYCL.  OpenMP's solution is tied to SIMD execution, and the concept of uniformity only exists when the compiler is performing loop or function vectorization.  The concept of uniformity in ISPC does more than state that the values are the same, and requests that  a single scalar mutable variable be shared between all SIMD lanes executing a vectorized program.

The concept of uniformity that we are proposing here is closer to the idea of uniformity in SPIR-V (see https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_uniformity_a_uniformity).  A variable is uniform if all work-items have the same value -- a uniform variable is not required to be stored in a scalar register, although implementations are free to employ this optimization if they wish.

I'm not saying that a user cannot shoot themselves in the foot by asserting that a value is uniform when it isn't, but this sort of thing exists in many places already.  If a user asserts that a pointer is aligned (using https://en.cppreference.com/w/cpp/memory/assume_aligned) and it isn't true at run-time, they will also get undefined behavior.

skozhukh replied:
A user will not shoot himself if we don't give them a gun that can allow such shooting. We at least should give a ways to check that it's really uniform.

Pennycook replied:
There is already a way to check that a variable has the same value of `x` for all work-items in some group `g`:
```c++
sycl::all_of_group(g, (x == sycl::group_broadcast(g, x)))
```

keryell replied:
It seems easy anyway to have `uniform` returning a wrapping type which is interpreted by `invoke_simd` to verify it is actually uniform at runtime with a debug mode of the runtime.
If at the end it doubles the performance on some hardware while having some checking mode possible, I can accept this SIMD foot gun inside SYCL. :-)
It is not worse than a C++ user using the wrong parallel policy in C++20 algorithms.

Pennycook replied:
> It seems easy anyway to have `uniform` returning a wrapping type which is interpreted by `invoke_simd` to verify it is actually uniform at runtime with a debug mode of the runtime.

Completely agree.  Nothing is stopping implementations from providing this behavior if they want, but since the performance overhead of checking every `uniform` declaration may be quite high, we shouldn't bake it in to the specification as a required default behavior.

skozhukh commented on sycl/doc/extensions/Uniform/Uniform.asciidoc:
````diff
@@ -0,0 +1,193 @@
+Asserting that all work-items in the sub-group access the same memory location:
+```c++
+float x = array[sycl::ext::oneapi::experimental::uniform(index)];
+```
````
Assuming that the uniformity should be processed by the compiler back end, what is your suggestion about IR produced by all the statements above where uniform is used..

Pennycook replied:
That's an implementation detail, and outside of the specification.

That said... remember that SYCL `uniform` does not imply that the variable must be stored differently, and so there is no need for a compiler to work backwards from a `uniform` declaration to examine all prior uses of a variable.  In fact, there is no need for a compiler to do anything at all based on a `uniform` declaration, and this is the expected implementation for some architectures (e.g. NVIDIA GPUs).

If a SYCL implementation wants to use a `uniform` declaration to decorate the variable with some special metadata that propagates to future uses of the variable, it is free to do so.  If it wants to work backwards from the declaration and prove (retroactively) that certain usages of the variable must also have been uniform, it is free to do that too.  But that is not required.

keryell reviewed (dismissed):
Nice to see no attribute or pragma used in this SYCL extension! :-)
I think this is the way we need to move SYCL forward on various hardware.

kbobrovs commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,368 @@
+pack. `Callable` may be a function object, a lambda,
+or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
+`Callable` must be an immutable callable with the same type and state for all
+work-items in the sub-group, otherwise behavior is undefined.
```
A late nit: I feel like some more explanation would be helpful here: if a callable is a lambda, it is so easy to unintendedly capture something non-uniform, which will then be used inside simd function. Suggest to add non-normative comment that this particular violation of  the `immutable callable with the same type and state for all work-items in the sub-group` principle should be watched out for.

Pennycook replied:
Thanks, Konst.  I've added a note in ba8ff0f711becf664b61c6dada597106da44e945.

kbobrovs replied:
Thanks John!

kbobrovs commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,373 @@
+To enable overloading in cases where the `Callable` accepts either a scalar
+type or a SIMD type, we leverage the `sycl::ext::oneapi::experimental::uniform` wrapper
+type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.
+
```
one more late nit: it would be good to have an example how overloaded callables should be passed to invoke_simd. E.g. this won't compile - scale1 is overloaded, so `scale1` as a function pointer is ambiguous (thanks to @skozhukh for bringing up). I guess there are general C++ tools for that, but since overloading is mentioned, it would help to see how it is supposed to be done with invoke_simd.

```cpp
float scale1(sycl::ext::oneapi::simd<float, 8> x, float n) {
  return x[2] * n;
}

float scale1(sycl::ext::oneapi::simd<float, 8> x, sycl::ext::oneapi::simd<float, 8> n) {
  return x[2] * n[2];
}

...
  float x = 1;
  float n = 2;
  // invoke SIMD function
  // x values from each work-item are combined into a simd<float, 8>
  auto y = sycl::ext::oneapi::invoke_simd(
      sg, scale1, x, sycl::ext::oneapi::uniform{n});

```

kbobrovs replied:
... like this:
```cpp
  auto y = sycl::ext::oneapi::invoke_simd<float(sycl::ext::oneapi::simd<float, 8>, float)>(
      sg, scale1, x, sycl::ext::oneapi::uniform{n});
```
?

Pennycook replied:
That would be one way to write it.  Anything that causes `scale1` to become a pointer to a specific function will work (see https://en.cppreference.com/w/cpp/language/overloaded_address).

So something like this would work, too:
```c++
auto f = static_cast<float(sycl::ext::oneapi::simd<float, 8>, float)>(scale1);
auto y = sycl::ext::oneapi::invoke_simd(sg, f, x, sycl::ext::oneapi::uniform{n});
```

There's also a neat trick where you can "lift" a function via a lambda, so that it would automatically call the correct version of `scale1` based on the arguments.  There's a good summary of that approach here: https://www.bfilipek.com/2019/03/lambdas-story-part2.html#bonus-lifting-with-lambdas.

This does make me think that the following line from the specification may be wrong:

> After this mapping has taken place, an appropriate variant of `Callable` is selected based on standard overload resolution.

@rolandschulz, what do you think about this?  It certainly looks to me like if the user passes the name of an overloaded function to `invoke_simd`, we have to resolve it _before_ we can call `invoke_simd`.  Is this a bug, or is there a trick I'm missing?

Pennycook replied:
I spoke to Roland offline about this.  The specification is correct, but we might want to consider adding a non-normative note and some examples to help people understand how this is expected to work.

The comment about overloading refers to situations where the `Callable` itself represents an overload set (e.g. because it is a functor with multiple call operators).  Your example doesn't fit this case, because what you're passing to `invoke_simd` is the name of an overloaded function (which decays to a function pointer).  It's the user's responsibility to handle that ambiguity.  Note that none of this is really specific to `invoke_simd`; it matches the behavior of functions like `std::invoke` and `std::apply`.

There's no right answer here, but we should adopt a consistent style for training examples.  If the user doesn't want `invoke_simd` to automatically select an overload, the clearest syntax is the one that you proposed:
```c++
  auto y = sycl::ext::oneapi::invoke_simd<float(sycl::ext::oneapi::simd<float, 8>, float)>(sg, scale1, x, sycl::ext::oneapi::uniform{n});
```

If the user does want `invoke_simd` to automatically select an overload, converting the functions into a functor with multiple call operators is the clearest syntax:
```c++
struct scale1 {
  float operator()(sycl::ext::oneapi::simd<float, 8> x, float n) {
    return x[2] * n;
  }

  float operator()(sycl::ext::oneapi::simd<float, 8> x, sycl::ext::oneapi::simd<float, 8> n) {
    return x[2] * n[2];
  }
}
...
  auto y = sycl::ext::oneapi::invoke_simd(sg, scale1(), x, sycl::ext::oneapi::uniform{n});
```
...although note that we now have to pass an instance of `scale1` to `invoke_simd` instead of just the name.

kbobrovs replied:
Thanks for the explanation, John.

> but we might want to consider adding a non-normative note and some examples to help people understand how this is expected to work

yes, please

Pennycook replied:
Added in https://github.com/intel/llvm/pull/3445/commits/b278fb495b8ce763667dc8cacfe1ebbc1c2b8d08.

kbobrovs commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,373 @@
+from the previous paragraph. This form of `invoke_simd` should be used with
+caution.
+
+Each argument to the invoked `Callable` must be a scalar type, a SIMD type,
```
Re-reading this again I discovered I might be missing something. Is this really what is meant?
```suggestion
Each formal parameter of the invoked `Callable` must be a scalar type, a SIMD type,
```
otherwise (if this is really about actual arguments passed to `invoke_simd`), it is not clear what types the corresponding formal parameter will (may) have, and which part of argument conversion rules apply in this case.

Pennycook replied:
I didn't appreciate the difference between "argument" and "parameter" until now.  You're right, this paragraph is talking about the types that appear in the function signature of the `Callable`.  I'll make a pass through the specification and make sure we're using these words correctly.

kbobrovs replied:
thanks, good to know formal parameters are meant

Pennycook replied:
I couldn't find any other places where we'd mixed them up, but I went ahead and corrected this one in https://github.com/intel/llvm/pull/3445/commits/a00319d39431b3ade7fea34a03931a4798a72f06.

kbobrovs commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,373 @@
+trivially copyable type wrapped in a `sycl::ext::oneapi::experimental::uniform`, or a
+`std::tuple` of these types.  Arguments may not be pointers or references,
+but pointers (like any other non-arithmetic type) may be passed if wrapped in a
+`sycl::ext::oneapi::experimental::uniform`.  Any such pointer value must point to memory
```
This basically forbids to have a simd callee with vector-of-pointers argument:
```cpp
void simd_foo(simd<int*,16> ptrs, ...) { scatter(ptrs,...); }
```
And the spec implies non-uniform pointer is represented as a base pointer + a vector of offsets, so SIMD library functions must be written with this style to be callable from SYCL:
```cpp
void simd_foo(int *base, simd<uintptr_t,16> offsets, ...) { ... }
```
This is natural for many platforms, but this implication should better be explicitly mentioned, I think.

(this and my other late comments can be addressed in future revisions of the spec, not a blocker for merge)

Pennycook replied:
I don't think this is a significant usability issue, because you can write a wrapper:
```c++
void simd_foo_from_sycl(int *base, simd<uintptr_t,16> offsets, ...) {
  simd<int*, 16> ptrs(base + offsets);
  simd_foo(ptrs, ...);
}
```

Would you prefer a non-normative note pointing out this is possible, or an unresolved issue suggesting we should revisit this restriction?

kbobrovs replied:
Ah, you are right, thanks. As simple as that :)

> Would you prefer a non-normative note pointing out this is possible, or an unresolved issue suggesting we should revisit this restriction?

Non-normative note would be enough.

Pennycook replied:
Added in https://github.com/intel/llvm/pull/3445/commits/b91567f38984c49f529b0ab2dc215fce83563f89.

kbobrovs commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,373 @@
+SPMD context of the calling kernel and the SIMD context of the callee,
+converting arguments and return values between scalar and SIMD types as
+appropriate.
+
```
Per offline discussion with @rolandschulz, the spec might need to require that the caller of `invoke_simd` should have explicitly set subgroup size - i.e. bear the `[[sycl::reqd_sub_group_size(N)]]` attribute. 

For given SPMD caller `foo` and SIMD callee `bar` the SG size could be deduced from vector arguments of the `bar` (unless `bar` signature contains uniform types only, in which case SG size does not matter), but that would present problems for SPMD callers of the `foo`, which must know if it has fixed SG size, esp. if the caller does not see `foo` body.

kbobrovs commented on sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc:
```diff
@@ -0,0 +1,440 @@
+from the previous paragraph. This form of `invoke_simd` should be used with
+caution.
+
+Each parameter of the invoked `Callable` must be a scalar type, a SIMD type,
```
```suggestion
Each formal parameter of the invoked `Callable` must be a scalar type, a SIMD type,
```
